File: exercises/java/src/test/java/workshop/exercise3/OrderProcessorTest.java
package workshop.exercise3;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import java.math.BigDecimal;
import static org.junit.jupiter.api.Assertions.*;

public class OrderProcessorTest {
    private OrderProcessor processor;

    @BeforeEach
    void setUp() {
        processor = new OrderProcessor();
    }

    @ParameterizedTest
    @CsvSource({
        "0.00, OH, Invalid Amount",
        "-10.00, AZ, Invalid Amount"
    })
    void processOrder_InvalidAmount_ReturnsFailure(String amountStr, String state, String expectedMsg) {
        BigDecimal amount = new BigDecimal(amountStr);
        OrderResult result = processor.processOrder(amount, state, false, 0);

        assertFalse(result.isSuccess());
        assertEquals(BigDecimal.ZERO, result.getFinalAmount());
        assertEquals(expectedMsg, result.getMessage());
    }

    @Test
    void processOrder_UnsupportedState_ReturnsFailure() {
        OrderResult result = processor.processOrder(new BigDecimal("100.00"), "NY", false, 0);

        assertFalse(result.isSuccess());
        assertEquals("Unsupported State", result.getMessage());
    }

    @ParameterizedTest
    @CsvSource({
        "100.00, OH, 105.75",
        "100.00, AZ, 105.60",
        "100.00, TX, 106.25"
    })
    void processOrder_ValidNonMember_CalculatesTax(String amountStr, String state, String expectedFinalStr) {
        BigDecimal amount = new BigDecimal(amountStr);
        BigDecimal expectedFinal = new BigDecimal(expectedFinalStr);

        OrderResult result = processor.processOrder(amount, state, false, 0);

        assertTrue(result.isSuccess());
        assertEquals(expectedFinal, result.getFinalAmount());
        assertEquals("Processed", result.getMessage());
    }

    @ParameterizedTest
    @CsvSource({
        "100.00, OH, 1, 103.75",
        "100.00, OH, 3, 100.75",
        "100.00, OH, 7, 90.75"
    })
    void processOrder_MemberDiscounts_AppliesCorrectTiers(
            String amountStr, String state, int loyaltyYears, String expectedFinalStr) {
        BigDecimal amount = new BigDecimal(amountStr);
        BigDecimal expectedFinal = new BigDecimal(expectedFinalStr);

        OrderResult result = processor.processOrder(amount, state, true, loyaltyYears);

        assertTrue(result.isSuccess());
        assertEquals(expectedFinal, result.getFinalAmount());
    }
}

File: exercises/javascript/exercise3/orderProcessor.test.js
const OrderProcessor = require('./orderProcessor');

describe('OrderProcessor', () => {
  let processor;

  beforeEach(() => {
    processor = new OrderProcessor();
  });

  test.each([
    [0, 'OH', 'Invalid Amount'],
    [-10, 'AZ', 'Invalid Amount'],
  ])('invalid amount (%p, %p) returns failure', (amount, state, expectedMsg) => {
    const result = processor.processOrder(amount, state, false, 0);
    expect(result.isSuccess).toBe(false);
    expect(result.finalAmount).toBe(0);
    expect(result.message).toBe(expectedMsg);
  });

  test('unsupported state returns failure', () => {
    const result = processor.processOrder(100.0, 'NY', false, 0);
    expect(result.isSuccess).toBe(false);
    expect(result.message).toBe('Unsupported State');
  });

  test.each([
    [100.0, 'OH', false, 0, 105.75],
    [100.0, 'AZ', false, 0, 105.60],
    [100.0, 'TX', false, 0, 106.25],
  ])('valid non-member order calculates tax correctly', (amount, state, isMember, loyaltyYears, expectedFinal) => {
    const result = processor.processOrder(amount, state, isMember, loyaltyYears);
    expect(result.isSuccess).toBe(true);
    expect(result.finalAmount).toBe(expectedFinal);
    expect(result.message).toBe('Processed');
  });

  test.each([
    [100.0, 'OH', true, 1, 103.75],
    [100.0, 'OH', true, 3, 100.75],
    [100.0, 'OH', true, 7, 90.75],
  ])('member discounts apply correct tiers', (amount, state, isMember, loyaltyYears, expectedFinal) => {
    const result = processor.processOrder(amount, state, isMember, loyaltyYears);
    expect(result.isSuccess).toBe(true);
    expect(result.finalAmount).toBe(expectedFinal);
  });
});
